### PR TITLE
Fixes for Ubuntu 16.04 and CMake 3.5

### PR DESCRIPTION
**Pull request by schornakj**
These changes are sort of contrary to the spirit of #3 , but they may be neded to support systems running Ubuntu 16.04 with CMake 3.5.

- Remove imported OpenMP targets, since [versions of CMake older than 3.9 don't generate them](https://stackoverflow.com/questions/12399422/how-to-set-linker-flags-for-openmp-in-cmakes-try-compile-function).
- Use old-style `include_directories()` function if a version of CMake older than 3.9 is detected.

### CODE REVIEW

File: yak/CMakeLists.txt [hunk 1]
```diff
@@ -73,6 +73,11 @@ set_directory_properties( PROPERTIES COMPILE_DEFINITIONS "" )
 link_directories(${PCL_LIBRARY_DIRS})
 add_definitions(${PCL_DEFINITIONS})
 
+if(${CMAKE_VERSION} VERSION_LESS "3.9.0")
+  message(WARNING "You are using CMake version ${CMAKE_VERSION}. Consider upgrading!")
+  include_directories(include)  # TODO: Avoid using include_directories at all!
+endif()
```
Comment (Levi-Armstrong) [inline yak/CMakeLists.txt:88]: Why is this required?

File: yak/CMakeLists.txt [hunk 2]
```diff
@@ -190,7 +194,7 @@ target_include_directories(${PROJECT_NAME}_marching_cubes PUBLIC
 target_link_libraries(${PROJECT_NAME}_marching_cubes
     PUBLIC
   ${PROJECT_NAME}
-  OpenMP::OpenMP_CXX
+  ${OpenMP_CXX_LIBRARIES}
```
Comment (Levi-Armstrong) [inline yak/CMakeLists.txt]: Recommend creating an imported target. This should also be added to your cmake config for this package.

```

find_package(OpenMP REQUIRED)

if(NOT TARGET OpenMP::OpenMP_CXX)
    find_package(Threads REQUIRED)
    add_library(OpenMP::OpenMP_CXX IMPORTED INTERFACE)
    set_property(TARGET OpenMP::OpenMP_CXX
                 PROPERTY INTERFACE_COMPILE_OPTIONS ${OpenMP_CXX_FLAGS})
    # Only works if the same flag is passed to the linker; use CMake 3.9+ otherwise (Intel, AppleClang)
    set_property(TARGET OpenMP::OpenMP_CXX
                 PROPERTY INTERFACE_LINK_LIBRARIES ${OpenMP_CXX_FLAGS} Threads::Threads)

endif()
```

File: yak/CMakeLists.txt (hunk 1 above)
Comment (schornakj) [reply yak/CMakeLists.txt:88]: When I build using CMake 3.5 it isn't able to find the header files in the include directory. I think there might be a problem with the generator expressions in the `target_include_directories` function.

File: yak/CMakeLists.txt (hunk 2 above)
Comment (schornakj) [reply yak/CMakeLists.txt]: I added the suggestion function to CMakeLists.txt. What should I add to the package CMake config?

File: yak/CMakeLists.txt (hunk 1 above)
Comment (schornakj) [reply yak/CMakeLists.txt:88]: This is the specific error:

```
Errors     << yak:make /home/joe/workspaces/yak_ws/logs/yak/build.make.003.log                                                                                                                                                        
/home/joe/workspaces/yak_ws/src/yak/yak/src/cuda/proj_icp.cu:1:39: fatal error: yak/kfusion/cuda/device.hpp: No such file or directory
compilation terminated.
/home/joe/workspaces/yak_ws/src/yak/yak/src/cuda/imgproc.cu:1:39: fatal error: yak/kfusion/cuda/device.hpp: No such file or directory
compilation terminated.
CMake Error at yak_generated_imgproc.cu.o.cmake:207 (message):
  Error generating
  /home/joe/workspaces/yak_ws/build/yak/CMakeFiles/yak.dir/src/cuda/./yak_generated_imgproc.cu.o


CMake Error at yak_generated_proj_icp.cu.o.cmake:207 (message):
  Error generating
  /home/joe/workspaces/yak_ws/build/yak/CMakeFiles/yak.dir/src/cuda/./yak_generated_proj_icp.cu.o


make[2]: *** [CMakeFiles/yak.dir/src/cuda/yak_generated_imgproc.cu.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [CMakeFiles/yak.dir/src/cuda/yak_generated_proj_icp.cu.o] Error 1
/home/joe/workspaces/yak_ws/src/yak/yak/src/cuda/tsdf_volume.cu:1:39: fatal error: yak/kfusion/cuda/device.hpp: No such file or directory
compilation terminated.
CMake Error at yak_generated_tsdf_volume.cu.o.cmake:207 (message):
  Error generating
  /home/joe/workspaces/yak_ws/build/yak/CMakeFiles/yak.dir/src/cuda/./yak_generated_tsdf_volume.cu.o


make[2]: *** [CMakeFiles/yak.dir/src/cuda/yak_generated_tsdf_volume.cu.o] Error 1
make[1]: *** [CMakeFiles/yak.dir/all] Error 2
make: *** [all] Error 2
cd /home/joe/workspaces/yak_ws/build/yak; catkin build --get-env yak | catkin env -si  /usr/bin/make --jobserver-fds=6,7 -j; cd -
```

Comment (schornakj) [reply yak/CMakeLists.txt:88]: I think I found the problem: the older-style CUDA support provided in the CMake 3.5 FindCUDA module doesn't support the `target_include_directories` directive ([similar issue here](https://gitlab.kitware.com/cmake/cmake/issues/18604), [brief summary here](https://codeyarns.com/2015/03/26/target_include_directories-does-not-work-with-cuda/), [original bug report here](https://cmake.org/Bug/view.php?id=14201)).

This problem has been fixed with the improvements to CUDA support added in newer versions of CMake (3.9 and later, I think). [This PR](https://github.com/ros-industrial/yak/pull/3) uses CMake 3.10 and the "modern CMake" approach, and `target_include_directories` works just fine there.

Little bit frustrating from the perspective of working towards a cross-compatible project...

**Comment by gavanderhoorn** [comment]
Approaching this differently: you may want to consider just requiring a more recent version of CMake than 3.5: KitWare has a (bw-compatible afaict) `apt` repository for Ubuntu (and Debian) users that ships the most recent versions of CMake for Xenial, Bionic, etc.

See [apt.kitware.com](https://apt.kitware.com/).

I'm currently running 3.15 on some Xenial machines. So far I've not run into any real issues, but I've not 'stressed' it too much.

That should allow you to use the same build scripts on Xenial and Bionic.

(whether this works is of course also dependent on the 'behaviour' of other CMake projects/packages, but that I cannot check for you)

**Comment by schornakj** [comment]
#3 enforces a minimum CMake version of 3.10 and provides instructions for how to install a newer version if needed (i.e. on Xenial). CMake is generally good about maintaining backwards compatibility so I think this is a decent solution.

In any case this PR is redundant and obsolete so I'm going to close it.
